python/dgl/nn/pytorch/explain/subgraphx.py

```diff
@@ -364,3 +365,453 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
+                include_value = include_probs[:, self.target_class]
+            marginal_contributions.append(include_value - exclude_value)
+
+        return torch.cat(marginal_contributions).mean().item()
+
+    def get_mcts_children(self, mcts_node):
+        r"""Get the children of the MCTS node for the search.
+
+        Parameters
+        ----------
+        mcts_node : MCTSNode
+            Node in MCTS
+
+        Returns
+        -------
+        list
+            Children nodes after pruning
+        """
+        if len(mcts_node.children) > 0:
+            return mcts_node.children
+
+        subg = node_subgraph(self.graph, mcts_node.nodes)
+        # Choose k nodes based on the highest degree in the subgraph
+        node_degrees_map = {
+            ntype: torch.zeros(
+                subg.num_nodes(ntype), device=subg.nodes(ntype).device
+            )
+            for ntype in subg.ntypes
+        }
+        for c_etype in subg.canonical_etypes:
+            src_ntype, _, dst_ntype = c_etype
+            node_degrees_map[src_ntype] += subg.out_degrees(etype=c_etype)
+            node_degrees_map[dst_ntype] += subg.in_degrees(etype=c_etype)
+
+        node_degrees_list = [
+            ((ntype, i), degree)
+            for ntype, node_degrees in node_degrees_map.items()
+            for i, degree in enumerate(node_degrees)
+        ]
+        node_degrees = torch.stack([v for _, v in node_degrees_list])
+        k = min(subg.num_nodes(), self.num_child)
+        chosen_node_indicies = torch.topk(
+            node_degrees, k, largest=self.high2low
+        ).indices
+        chosen_nodes = [node_degrees_list[i][0] for i in chosen_node_indicies]
+
+        mcts_children_maps = dict()
+
+        for ntype, node in chosen_nodes:
+            new_subg = remove_nodes(subg, node, ntype, store_ids=True)
+
+            if new_subg.num_edges() > 0:
+                new_subg_homo = to_homogeneous(new_subg)
+                # Get the largest weakly connected component in the subgraph.
+                nx_graph = to_networkx(new_subg_homo.cpu())
+                largest_cc_nids = list(
+                    max(nx.weakly_connected_components(nx_graph), key=len)
+                )
+                largest_cc_homo = node_subgraph(new_subg_homo, largest_cc_nids)
+                largest_cc_hetero = to_heterogeneous(
+                    largest_cc_homo, new_subg.ntypes, new_subg.etypes
+                )
+
+                # Follow steps for backtracking to original graph node ids
+                # 1. retrieve instanced homograph from connected-component homograph
+                # 2. retrieve instanced heterograph from instanced homograph
+                # 3. retrieve hetero-subgraph from instanced heterograph
+                # 4. retrieve orignal graph ids from subgraph node ids
+                cc_nodes = {
+                    ntype: subg.ndata[NID][ntype][
+                        new_subg.ndata[NID][ntype][
```

Comment:
This may not be correct. Use dgl.NTYPE to also backtrack node types.

Reply:
I believe the `dgl.NTYPE` is used already when using `to_heterogeneous`?  
but here is my thought process on the matter as I'm not yet familiar with the internals if anyone has a cleaner solution:

> example scenario is from MUTAG dataset

say there is an initial graph like the following:

```
{
    'C': tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14]),
    'N': tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]),
    'O': tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17])
}
```

say the current MCTS exploration subgraph looks like so:

```
{
    'C': tensor([ 1,  2,  3,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14]),
    'N': tensor([15]),
    'O': tensor([16, 17])
}
```

the relative mapping shows up in dgl as the following:

```
{
    '_ID': {
        'C': tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12]),
        'N': tensor([0]),
        'O': tensor([0, 1])
    }
}
```

In the process of generating more MCTS candidates, we remove `C[7]` (which is really `C-9` of the original graph) giving us a new subgraph instance:

```
{
    '_ID': {
        'C': tensor([ 0,  1,  2,  3,  4,  5,  6,  8,  9, 10, 11, 12]),
        'N': tensor([0]),
        'O': tensor([0, 1])
    }
}
```

From here we convert into a homograph and compute the largest connected components using networkX. Then we obtain the nods in a heterograph format.

```
# dgl homograph
{
    '_ID': tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11,  0,  0,  1]),
    '_TYPE': tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 2])
}

# networkX, resulting connected component nids:
[7, 8, 9, 10, 11, 12, 13, 14]

# dgl.subgraph using above nids mask
{
    '_ID': tensor([ 7,  8,  9, 10, 11, 12, 13, 14]),
    '_TYPE': tensor([0, 0, 0, 0, 0, 1, 2, 2])
}

# dgl.to_heterogenous using above graph
{
    '_ID': {
        'C': tensor([0, 1, 2, 3, 4]),
        'N': tensor([5]),
        'O': tensor([6, 7])
    }
}
```

The new MCTS node I would expect looks like so,

```
{
    'C': tensor([10, 11, 12, 13, 14]),
    'N': tensor([15]),
    'O': tensor([16, 17])
}
```

which I was able to accomplish using the code block we are commenting on.

Here is a visual of this graph with labeled nodes if this helps:
![test_MUTAG_explain](https://user-images.githubusercontent.com/48701178/232169418-c557c60c-b38f-4f4a-84f8-f6a0c4c83b15.png)

Reply:
I see. I think you are right. Thanks.